proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthTokenStore.java

```diff
@@ -0,0 +1,421 @@
+     * the same token while the task is in processing. You can use {@code PowerAuthTokenStore.hasLocalToken()} method
+     * to check, whether the token is already in the local database.
+     *
+     * @param context Context
+     * @param tokenName Name of requested token.
+     * @param authentication An authentication instance specifying what factors should be used for token creation.
+     * @param listener Listener with callbacks to receive a token.
+     * @return {@code AsyncTask} associated with the running server request or null if request has been processed synchronously.
+     */
+    public @Nullable AsyncTask requestAccessToken(@NonNull final Context context, @NonNull final String tokenName, @NonNull PowerAuthAuthentication authentication, @NonNull final IGetTokenListener listener) {
+
+        Throwable error = null;
+        PowerAuthPrivateTokenData tokenData = null;
+
+        synchronized (this) {
+            if (this.canRequestForAccessToken()) {
+                tokenData = this.getTokenData(context, tokenName);
+            } else {
+                error = new PowerAuthErrorException(PowerAuthErrorCodes.PA2ErrorCodeMissingActivation);
+            }
+        }
+
+        // If there's private data or error available, then report that immediately to the listener.
+        if (error != null) {
+            listener.onGetTokenFailed(error);
+            return null;
+        } else if (tokenData != null) {
+            listener.onGetTokenSucceeded(new PowerAuthToken(this, tokenData));
+            return null;
+        }
+        // Launch HTTP request
+
+        // 1) Encrypt empty data to get ephemeral key
+        final Pair<ECIESEncryptor, ECIESCryptogram> pair = encryptor.encryptRequestSynchronized(null);
+        if (pair == null) {
+            listener.onGetTokenFailed(new PowerAuthErrorException(PowerAuthErrorCodes.PA2ErrorCodeEncryptionError));
+            return null;
+        }
+        final ECIESEncryptor decryptor = pair.first;
+        final ECIESCryptogram cryptogram = pair.second;
+
+        // 2) Build post data
+        final TokenCreateRequest request = new TokenCreateRequest();
+        request.setEphemeralPublicKey(cryptogram.getKeyBase64());
+        final Pair<byte[], String> postData = httpClient.serializeRequestObject(request);
+
+        // 3) Sign that post data
+        final PowerAuthAuthorizationHttpHeader authHeader = sdk.requestSignatureWithAuthentication(context, authentication, "POST", PA2CreateTokenEndpoint.CREATE_TOKEN, postData.first);
+        if (!authHeader.isValid()) {
+            listener.onGetTokenFailed(new PowerAuthErrorException(authHeader.powerAuthErrorCode));
+            return null;
+        }
+        final HashMap<String, String> headers = new HashMap<>();
+        headers.put(authHeader.getKey(), authHeader.getValue());
+
+        // 4) Execute HTTP request
+        return httpClient.createToken(headers, request, new INetworkResponseListener<TokenCreateResponse>() {
```

Comment:
Note: You were right our networking could be greatly improved here. The signature header computation basically relies on the same serialization routine (object -> bytes) working the same way twice...